### PR TITLE
Add Kubernetes IPAM mode

### PR DESCRIPTION
**Pull request by tgraf**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10407)

### CODE REVIEW

File: Documentation/concepts/ipam/kubernetes.rst [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+=================================== ============================================================
+Annotation                          Description
+=================================== ============================================================
+``io.cilium.network.ipv4-pod-cidr`` IPv4 PodCIDR range
+``io.cilium.network.ipv6-pod-cidr`` IPv6 PodCIDR range
+=================================== ============================================================
+
+.. note:: The annotation-based mechanism is primiarly useful in combination with
+	  older Kubernetes versions which do not support ``Spec.PodCIDRs`` yet
+	  but support for both IPv4 and IPv6 is enabled.
```
Comment (aanm) [inline Documentation/concepts/ipam/kubernetes.rst:43]: Why are we adding this if we are removing the node annotation mode in 1.8? Will this still work?

**Review by aanm — CHANGES_REQUESTED** [review]
Also, shouldn't we add something in the upgrade guide with regarding the default value of the `ipam` option?

File: Documentation/concepts/ipam/kubernetes.rst (hunk 1 above)
Comment (tgraf) [reply Documentation/concepts/ipam/kubernetes.rst:43]: Yes, the annotations can still be set manually by the user. I think it's a good fallback way for k8s versions which do not supports `PodCIDRs`. Thoughts?

**Comment by tgraf** [comment]
> Also, shouldn't we add something in the upgrade guide with regarding the default value of the `ipam` option?

Does anything change for the user? This should be invisible except for the help text and the logfile which now documents that the legacy hostscope is in use.

**Comment by tgraf** [comment]
test-me-please

File: Documentation/concepts/ipam/kubernetes.rst (hunk 1 above)
Comment (aanm) [reply Documentation/concepts/ipam/kubernetes.rst:43]: But how? we don't have a node watcher in cilium-agent AFAIK.

Comment (tgraf) [reply Documentation/concepts/ipam/kubernetes.rst:43]: `waitForNodeInformation()` will `GET` the Kubernetes node and wait for the PodCIDR to be populated by either `PodCIDR`, `PodCIDRs` field or annotations.

Comment (aanm) [reply Documentation/concepts/ipam/kubernetes.rst:43]: oh that makes more sense.

**Review by aanm — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**